src/drivers/dshot/dshot.cpp

```diff
@@ -0,0 +1,1678 @@
+		}
+
+	} else {
+		PX4_ERR("alloc failed");
+	}
+
+	delete instance;
+	_object.store(nullptr);
+	_task_id = -1;
+
+	return PX4_ERROR;
+}
+
+void
+DShotOutput::capture_trampoline(void *context, uint32_t chan_index,
+				hrt_abstime edge_time, uint32_t edge_state, uint32_t overflow)
+{
+	DShotOutput *dev = reinterpret_cast<DShotOutput *>(context);
+	dev->capture_callback(chan_index, edge_time, edge_state, overflow);
+}
+
+void
+DShotOutput::capture_callback(uint32_t chan_index,
+			      hrt_abstime edge_time, uint32_t edge_state, uint32_t overflow)
+{
+	fprintf(stdout, "DShot: Capture chan:%d time:%lld state:%d overflow:%d\n", chan_index, edge_time, edge_state, overflow);
+}
+
+void
+DShotOutput::update_dshot_out_state(bool on)
+{
+	if (on && !_outputs_initialized && _output_mask != 0) {
+		DShotConfig config = (DShotConfig)_param_dshot_config.get();
+		unsigned dshot_frequency;
+
+		switch (config) {
+		case DShotConfig::DShot150:
+			dshot_frequency = DSHOT150;
+			break;
+
+		case DShotConfig::DShot300:
+			dshot_frequency = DSHOT300;
+			break;
+
+		case DShotConfig::DShot1200:
+			dshot_frequency = DSHOT1200;
+			break;
+
+		case DShotConfig::DShot600:
+		default:
+			dshot_frequency = DSHOT600;
+			break;
+		}
+
+		int ret = up_dshot_init(_output_mask, dshot_frequency);
+
+		if (ret != 0) {
+			PX4_ERR("up_dshot_init failed (%i)", ret);
+			return;
+		}
+
+		_outputs_initialized = true;
+	}
+
+	if (_outputs_initialized) {
+		up_dshot_arm(on);
+		_outputs_on = on;
+	}
+}
+
+void DShotOutput::updateTelemetryNumMotors()
+{
+	if (!_telemetry) {
+		return;
+	}
+
+	int motor_count = 0;
+
+	if (_mixing_output.mixers()) {
+		motor_count = _mixing_output.mixers()->get_multirotor_count();
+	}
+
+	_telemetry->handler.setNumMotors(motor_count);
+}
+
+void DShotOutput::initTelemetry(const char *device)
+{
+	if (!_telemetry) {
+		_telemetry = new Telemetry{};
+
+		if (!_telemetry) {
+			PX4_ERR("alloc failed");
+			return;
+		}
+	}
+
+	int ret = _telemetry->handler.init(device);
+
+	if (ret != 0) {
+		PX4_ERR("telemetry init failed (%i)", ret);
+	}
+
+	updateTelemetryNumMotors();
+}
+
+void DShotOutput::handleNewTelemetryData(int motor_index, const DShotTelemetry::EscData &data)
+{
+	// fill in new motor data
+	esc_status_s &esc_status = _telemetry->esc_status_pub.get();
+
+	if (motor_index < esc_status_s::CONNECTED_ESC_MAX) {
+		esc_status.esc_online_flags |= 1 << motor_index;
+		esc_status.esc[motor_index].timestamp = data.time;
+		esc_status.esc[motor_index].esc_rpm = ((int)data.erpm * 100) / (_param_mot_pole_count.get() / 2);
+		esc_status.esc[motor_index].esc_voltage = (float)data.voltage * 0.01f;
+		esc_status.esc[motor_index].esc_current = (float)data.current * 0.01f;
+		esc_status.esc[motor_index].esc_temperature = data.temperature;
+		// TODO: accumulate consumption and use for battery estimation
+	}
+
+	// publish when motor index wraps (which is robust against motor timeouts)
+	if (motor_index < _telemetry->last_motor_index) {
```

Comment:
When we were testing dshot with just one ESC connected we could not get telemety info on `esc_status` because of this line. I don't know if 1 ESC is an actual use case (planes?), but if it is, would the following be better?

```
if (motor_index <= _telemetry->last_motor_index) {
```

Reply:
Good point. I don't see any issues with that. Can you do a PR?